### PR TITLE
Support bnb l2

### PR DESCRIPTION
**Pull request by boyuan-chen**
:clipboard: Add associated issues, tickets, docs URL here.

## Overview

Describe what your Pull Request is about in a few sentences.

Add git patch to support BNB L2

## Changes

Describe your changes and implementation choices. More details make PRs easier to review.

#### BNB
Currently, there are two ways to deposit the L1 native token (BNB) to L2. Users can use `OptimismPortal` or `L1StandardBridge` to deposit the L1 native token. To simplify things, I disabled the L1 native token deposit in `L1StandardBridge`. Users **have to** use OptimismPortal to deposit BNB.

In `OptimismPortal`, users can force L2 to include transactions from L1. However, the deposit of the L1 native token will mint an L2 ERC20 token on L2. I separated this native token deposit from the original logic and added another event to mint the ERC20 token on L2.

#### BOBA
I added a `depositBoba` function in `OptimismPortal` to emit an event so that we can mint the L2 native token on L2. For the security consideration, this function can **only** be called from `L1CrossDomainMessenger` and the initial sender **must** be `L1StandardBridge`. In other words, users **must** call `L1StandardBridge` to deposit BOBA.  

Workflow: `L1StandardBridge` -> `L1CrossDomainMessenger` -> `OptimismPortal`

#### BOBA
Users call the `withdrawETH` or `withdrawETHTo` functions in `L2StandardBridge`. The `L2StandardBridge` will call the `withdrawBoba` function in `L2CrossDomainMessenger `. The difference of the `withdrawBoba` and the standard `sendMessage` function is that we override the value in the `relayMessage` function, because we treat Boba as an ERC20 token on L1. The `withdrawBoba ` function calls the `initiateBobaWithdrawal` function in `L2ToL1MessagePasser` to emit the final event.

During this process, the BOBA token will be transferred from the user account to `L2ToL1MessagePasser` as the standard process and we can destroy the BOBA token in `L2ToL1MessagePasser `.

#### BNB
Users call the `withdrawERC20` or `withdrawERC20To` functions in `L2StandardBridge`. The `L2StandardBridge` will call the `withdrawBnb` function in `L2CrossDomainMessenger`. The difference of the `withdrawBnb` and the standard `sendMessage` function is that we override the value in the `relayMessage` function, because we treat Bob as the L1 native token on L1. The `withdrawBnb` function calls the `initiateBnbWithdrawal` function in `L2ToL1MessagePasser` to emit the final event.

## Testing

Describe how to test your new feature/bug fix and if possible, a step by step guide on how to demo this.

To start the network, run this command:
```
make devnet-bnb-hardhat-up
```
To run the deposit and withdrawal tests:
```
make devnet-bnb-hardhat-test
```

### CODE REVIEW

File: packages/sdk/tasks/deposit-bnb.ts [hunk 1]
```diff
@@ -0,0 +1,432 @@
...
+        console.log(
+          `Block hash changed from ${prevBlockHash} to ${messageReceipt.blockHash}`
+        )
+        i = 0
+      } else if (prevBlockHash !== '') {
+        console.log(`No reorg detected: ${i}`)
+      }
+
+      prevBlockHash = messageReceipt.blockHash
+      await sleep(1000)
```
Comment (jyellick) [inline packages/sdk/tasks/deposit-bnb.ts:194]: So, we always fetch the receipt 12 times, waiting 12 seconds in total, regardless of the success, failure, or re-org.  Is it intentional that we don't abort with an error if there is no success after all the tries?

File: packages/sdk/tasks/deposit-bnb.ts [hunk 2]
```diff
@@ -0,0 +1,432 @@
...
+    }
+    console.log(`Deposit BOBA confirmed`)
+
+    const l2BobaBalance = await l2Provider.getBalance(l2Signer.address)
+    if (l2BobaBalance.lt(utils.parseEther('10'))) {
```
Comment (jyellick) [inline packages/sdk/tasks/deposit-bnb.ts:199]: I'm no typescript expert, but this seems like it's asserting that the balance is simply non-zero -- shouldn't we be checking for an explicit balance?

File: packages/sdk/tasks/deposit-bnb.ts [hunk 3]
```diff
@@ -0,0 +1,432 @@
...
+      throw new Error(
+        `bad deposit. recipient balance on L2: ${utils.formatEther(
+          l2BobaBalance
+        )}`
+      )
+    }
+    console.log(`Deposit BOBA success`)
```
Comment (jyellick) [inline packages/sdk/tasks/deposit-bnb.ts:206]: Related to the above, but shouldn't we check both account balances and confirm they're as expected?

File: packages/sdk/tasks/deposit-bnb.ts [hunk 4]
```diff
@@ -0,0 +1,432 @@
...
+
+    const l2BnbBalance = await L2BnbToken.balanceOf(l2Signer.address)
+    if (l2BnbBalance.lt(utils.parseEther('1'))) {
+      throw new Error(
+        `bad deposit. recipient balance on L2: ${utils.formatEther(
+          l2BnbBalance
+        )}`
+      )
+    }
+    console.log(`Deposit BNB success`)
```
Comment (jyellick) [inline packages/sdk/tasks/deposit-bnb.ts:249]: Same comments here, we don't seem to catch if the transaction truly fails, and we don't look for appropriate balances.

File: packages/sdk/tasks/deposit-bnb.ts [hunk 5]
```diff
@@ -0,0 +1,432 @@
...
+      OptimismPortal.address
+    )
+
+    const expectedopBalanceFinally = opBalanceFinally.add(utils.parseEther('1'))
+    if (!expectedopBalanceFinally.eq(opBalanceAfter)) {
+      throw new Error(
+        `BNB Balance mismatch, expected: ${expectedopBalanceFinally}, actual: ${opBalanceAfter}`
+      )
+    }
+    console.log('BNB Withdrawal success')
```
Comment (jyellick) [inline packages/sdk/tasks/deposit-bnb.ts:431]: I'm not seeing checks that the withdraw is reflected both on L1 and L2 as far as amounts transferred.  Probably also worth validating that the other asset (ie BNB/ETH or Boba) is not changed.

**Review by jyellick — COMMENTED** [review]
While reading to try to make sure that I understood the protocol level interactions of the contracts, I stumbled across: https://specs.optimism.io/protocol/granite/custom-gas-token.html

It looks like this path may be officially supported soon in the specs?  Even if we don't want to wait for official support, I think it's worth investigating and maybe building from this experimental branch?

**Comment by boyuan-chen** [comment]
> While reading to try to make sure that I understood the protocol level interactions of the contracts, I stumbled across: https://specs.optimism.io/protocol/granite/custom-gas-token.html
> 
> It looks like this path may be officially supported soon in the specs? Even if we don't want to wait for official support, I think it's worth investigating and maybe building from this experimental branch?

Which branch are you referring to?

**Comment by boyuan-chen** [comment]
I found this branch. Optimism blocks some withdrawal paths to simplify the code changes. Basically, you can only withdraw your L2 native token from `L2ToL1MessagePasser`. We can simplify our codebase and use some functions from Optimism, but the idea is the same. One thing we can simplify a lot is that our L1 native token on L2 will always be `0x42000...23`, so we can hardcode this value instead of bypassing it from L1.
One different design is that Optimism override the L2 to L1 messages in `OptimismPortal`, but I send all correct messages from L2 to L1 so that we don't modify the logic of `OptimismPortal`.
https://github.com/ethereum-optimism/optimism/tree/feat/custom-gas-token

**Comment by jyellick** [comment]
> > While reading to try to make sure that I understood the protocol level interactions of the contracts, I stumbled across: https://specs.optimism.io/protocol/granite/custom-gas-token.html
> > It looks like this path may be officially supported soon in the specs? Even if we don't want to wait for official support, I think it's worth investigating and maybe building from this experimental branch?
> 
> Which branch are you referring to?

The specification referenced above has begun to be implemented in the current optimism tree.  For instance here:

https://github.com/ethereum-optimism/optimism/blob/943ed0f94f02a2ebad5155334828cc07fca46e0f/packages/contracts-bedrock/src/libraries/GasPayingToken.sol

You can see some of the implementation referenced in the protocol specs.

**Comment by boyuan-chen** [comment]
> > > While reading to try to make sure that I understood the protocol level interactions of the contracts, I stumbled across: https://specs.optimism.io/protocol/granite/custom-gas-token.html
> > > It looks like this path may be officially supported soon in the specs? Even if we don't want to wait for official support, I think it's worth investigating and maybe building from this experimental branch?
> > 
> > 
> > Which branch are you referring to?
> 
> The specification referenced above has begun to be implemented in the current optimism tree. For instance here:
> 
> https://github.com/ethereum-optimism/optimism/blob/943ed0f94f02a2ebad5155334828cc07fca46e0f/packages/contracts-bedrock/src/libraries/GasPayingToken.sol
> 
> You can see some of the implementation referenced in the protocol specs.

Yes! Let's merge the upstream, so we can use the same codebase.

File: 0001-BNB-L2.patch
Comment (souradeep-das) [inline 0001-BNB-L2.patch:1]: The `depositBoba` method on the L1CDM and the `withdrawBoba` and `withdrawBnb` methods on the L2CDM seem to allow calls from anyone, allowing to bypass the token deposits but send msg directly. The assumption of the call to initiate from StandardBridge -> CDM -> OptimismPortal doesn't seem to hold. could you verify?

File: 0001-BNB-L2.patch [hunk 6]
```diff
@@ -0,0 +1,835 @@
...
++        bytes memory opaqueData = abi.encodePacked(_value, _value, _gasLimit, false, _data);
+ 
+         // Emit a TransactionDeposited event so that the rollup node can derive a deposit
+         // transaction for this deposit.
+@@ -430,3 +501,37 @@ contract OptimismPortal is Initializable, ResourceMetering, ISemver {
+         return block.timestamp > _timestamp + l2Oracle.FINALIZATION_PERIOD_SECONDS();
+     }
+ }
++
++pragma solidity 0.8.15;
```
Comment (souradeep-das) [inline 0001-BNB-L2.patch:280]: minor: we could remove this second declaration

File: 0001-BNB-L2.patch
Comment (boyuan-chen) [reply 0001-BNB-L2.patch:1]: `depositBoba`, `withdrawBoba`, and `withdrawBnb` are the public-facing functions. This PR will be closed and reapplied in our develop branch

**Pull request closed without merging**